File: src/lib/contratos/parcelas.ts
// Porte de _docs/referencia-skill/scripts/parcelas.py — gera a tabela de N
// parcelas (Cláusula Oitava) a partir do 1º vencimento.
//
// Cuidado replicado do Python (dateutil.relativedelta): "somar 1 mês" gruda
// no dia grampeado ao último dia do mês de destino quando ele não existe
// (ex.: 31/01 + 1 mês = 28/02, não "03/03" como o overflow nativo do
// `Date.setMonth` do JS faria) — addMesesComGrampo() abaixo replica isso.

export type ConvencaoParcelas = "calendario" | "ciclo";

export interface Parcela {
  numero: number;
  mesVigencia: string;
  periodo: string;
  vencimento: string;
}

const MESES = [
  "",
  "Janeiro",
  "Fevereiro",
  "Março",
  "Abril",
  "Maio",
  "Junho",
  "Julho",
  "Agosto",
  "Setembro",
  "Outubro",
  "Novembro",
  "Dezembro",
];

function diasNoMes(ano: number, mesIndexUm: number): number {
  // mesIndexUm: 1-12. Dia 0 do mês seguinte = último dia do mês pedido.
  return new Date(ano, mesIndexUm, 0).getDate();
}

function addMesesComGrampo(data: Date, meses: number): Date {
  const dia = data.getDate();
  const totalMeses = data.getMonth() + meses; // getMonth() é 0-11
  const ano = data.getFullYear() + Math.floor(totalMeses / 12);
  const mesIndexUm = (((totalMeses % 12) + 12) % 12) + 1; // 1-12
  const diaGrampado = Math.min(dia, diasNoMes(ano, mesIndexUm));
  return new Date(ano, mesIndexUm - 1, diaGrampado);
}

function subtrairDias(data: Date, dias: number): Date {
  const resultado = new Date(data);
  resultado.setDate(resultado.getDate() - dias);
  return resultado;
}

function formatarData(data: Date): string {
  const dd = String(data.getDate()).padStart(2, "0");
  const mm = String(data.getMonth() + 1).padStart(2, "0");
  return `${dd}/${mm}/${data.getFullYear()}`;
}

/** primeiroVencimento no formato DD/MM/AAAA, igual ao script original. */
export function gerarParcelas(
  primeiroVencimento: string,
  n = 12,
  convencao: ConvencaoParcelas = "calendario",
): Parcela[] {
  const [dd, mm, aaaa] = primeiroVencimento.split("/").map(Number);
  let venc = new Date(aaaa, mm - 1, dd);
  const linhas: Parcela[] = [];

  for (let i = 0; i < n; i++) {
    let ini: Date;
    let fim: Date;
    if (convencao === "calendario") {
      ini = new Date(venc.getFullYear(), venc.getMonth(), 1);
      fim = new Date(venc.getFullYear(), venc.getMonth() + 1, 0); // último dia do mês
    } else {
      ini = venc;
      fim = subtrairDias(addMesesComGrampo(venc, 1), 1);
    }

    linhas.push({
      numero: i + 1,
      mesVigencia: `${MESES[venc.getMonth() + 1]}/${venc.getFullYear()}`,
      periodo: `${formatarData(ini)} a ${formatarData(fim)}`,
      vencimento: formatarData(venc),
    });

    venc = addMesesComGrampo(venc, 1);
  }

  return linhas;
}

File: _skill/scripts/parcelas.py
#!/usr/bin/env python3
"""
Gera a tabela de 12 parcelas (Cláusula Oitava) a partir do 1º vencimento.

Uso: python parcelas.py DD/MM/AAAA [n_parcelas] [dia_venc]

Reproduz o padrão dos contratos: cada parcela = 1 mês de vigência, período do dia
do vencimento ao dia anterior do mês seguinte, vencimento no dia fixo.
Imprime linhas "Parcela | Mês de Vigência | Período | Vencimento".
"""
import sys
from datetime import date
from dateutil.relativedelta import relativedelta

MESES = ["", "Janeiro", "Fevereiro", "Março", "Abril", "Maio", "Junho",
         "Julho", "Agosto", "Setembro", "Outubro", "Novembro", "Dezembro"]


def gerar(primeiro_venc, n=12, periodo="calendario"):
    """
    periodo="calendario" (default, padrão Goiás/2026): período = mês-calendário cheio
        (01/05/2026 a 31/05/2026), vencimento no dia informado.
    periodo="ciclo" (padrão Corinthians): período = do dia do vencimento ao dia anterior
        do mês seguinte (10/05/2026 a 09/06/2026).
    """
    d, m, a = map(int, primeiro_venc.split("/"))
    venc = date(a, m, d)
    linhas = []
    for i in range(n):
        if periodo == "calendario":
            ini = venc.replace(day=1)
            fim = ini + relativedelta(months=1) - relativedelta(days=1)
        else:  # ciclo
            ini = venc
            fim = venc + relativedelta(months=1) - relativedelta(days=1)
        vig = f"{MESES[venc.month]}/{venc.year}"
        per = f"{ini.strftime('%d/%m/%Y')} a {fim.strftime('%d/%m/%Y')}"
        linhas.append((i + 1, vig, per, venc.strftime("%d/%m/%Y")))
        venc = venc + relativedelta(months=1)
    return linhas


if __name__ == "__main__":
    if len(sys.argv) < 2:
        raise SystemExit("uso: python parcelas.py DD/MM/AAAA [n]")
    n = int(sys.argv[2]) if len(sys.argv) > 2 else 12
    modo = sys.argv[3] if len(sys.argv) > 3 else "calendario"
    for p in gerar(sys.argv[1], n, modo):
        print(f"{p[0]:>2} | {p[1]:<16} | {p[2]} | {p[3]}")
